### ISSUE TITLE
Healthchek support for containers

### ISSUE THREAD

**Issue by willianpaixao** [ethereum/go-ethereum#22442]
# Rationale

### Why should this feature exist?
To my knowledge, there's no "dead simple" way to know if your geth process is running smooth, syncing, or maybe even frozen.
I know you can send a RPC request asking for the `eth_syncing` object, but in this issue, I'd like to share the concept of liveness and readiness of a process. A container called "live" when its underlying process is running (aka not finished, crashed, or frozen), and it's called "ready" when is literally ready to receive requests (in this case is synced).  

### What are the use-cases?
Anyone running a monitoring tool could benefit from this feature, but especially the ones using a container solution, like Docker and/or Kubernetes. Modern container tools can monitor the state of a container using health checks. [[1]](https://docs.docker.com/engine/reference/builder/#healthcheck)
Once the health check is implemented, the orchestrating tool can spot, flag, and automatically restart your rouge container. [[2]](https://cloud.ibm.com/docs/go?topic=go-go-healthcheck)

# Implementation

### Do you have ideas regarding the implementation of this feature?
Yes, in theory, we add one endpoint to signal that our process is running with no problems, that returns HTTP 200. An additional feature would be another endpoint that would indicate an error, maybe syncing state.
The implementation can be with or without using a library. [[3]](https://github.com/docker/go-healthcheck) [[4]](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)

**Comment by ligi** [comment]
We see the value in it - unfortunately we do not see a way to implement a health-check that has a reliable output. There are cases (e.g. compatction) where false positives could occur. And then it is better not to have such a endpoint at all.

**Comment by MariusVanDerWijden** [comment]
You can always use a trivial endpoint like `eth_version` to see if your node is still responding to requests.
It's not easy to determine if your node is working "correctly" for example if your node has not imported a new block in 5 minutes it might either be that your node crashed, your network is down or that no-one mined a block in five minutes out of pure chance.

**Comment by willianpaixao** [comment]
Hey @ligi @karalabe,

Just a question, after Geth v.1.10.x and the release of Snap sync, is this issue possible to be implemented? Has anything changed to allow this feature?

Thanks in advance!

**Comment by willianpaixao** [comment]
Hi again @ligi @MariusVanDerWijden @karalabe, any updates?
I can implement the feature, but need a confirmation that it's desired.

**Comment by willianpaixao** [comment]
@holiman @ligi @MariusVanDerWijden I'm not sure if notifications are sent after an issue is closed.

Things have changed since this issue was closed (details in the comments above) can we implement this now?

**Comment by holiman** [comment]
>  implement this now?

What specifically is it you want to implement, that isn't covered already by `eth_syncing` and/or `eth_version` ?

**Comment by willianpaixao** [comment]
@holiman sorry I thought it was clear in the initial issue description. I'll try to summarize and maybe rephrase.

I am suggesting the addition of two entrypoints (preferably two http endpoints, but could be a geth cli combo as well). The first entrypoint is the simplest, it only states that the process is responsive (aka not frozen or even dead).

The second is the one that is tricky and started this discussion, it returns whether the geth process is ready to receive requests (aka is not syncing, or doing some blocking task).

As stated in the initial description, these entrypoints will be used by tools that will monitor the state of the process or container.
This would be specially useful for those orchestrating tools to know when geth is ready to receive requests once a new node is spun. [[1]](https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-setting-up-health-checks-with-readiness-and-liveness-probes)

**Comment by willianpaixao** [comment]
In my opinion yes, it should be enough.

But @MariusVanDerWijden and @ligi raised the question where corner cases could lead to false positives.
